### PR TITLE
staticanalysis: Display custom errors with codes

### LINKED ISSUE #1378: Categorize and report the errors
Every exception caught should be categorized and reported through Taskcluster indexed data.

### PR DESCRIPTION
**Pull request by La0**
Fixes #1378 , #1379 

These `error_code` will then be used by `staticanalysis/frontend` to display stats about error sources

Also display phabricator & bugzilla details

### CODE REVIEW

**Comment by marco-c** [comment]
What is the advantage over Sentry?

File: src/staticanalysis/bot/static_analysis_bot/workflow.py [hunk 1]
```diff
@@ -178,22 +185,24 @@ def run(self, revision):
...
                 logger.info('No clang or java files detected, skipping mach, infer and clang-*')
 
             # Setup python environment
             logger.info('Mach lint setup...')
             cmd = ['gecko-env', './mach', 'lint', '--list']
             out = run_check(cmd, cwd=settings.repo_dir)
-            assert 'error: problem with lint setup' not in out.decode('utf-8'), \
-                'Mach lint setup failed'
+            if 'error: problem with lint setup' in out.decode('utf-8'):
+                raise AnalysisException('mach', 'Mach lint setup failed')
```
Comment (marco-c) [inline src/staticanalysis/bot/static_analysis_bot/workflow.py]: Maybe use different codes for the different mach commands? This way you can have more detailed statistics.

**Comment by marco-c** [comment]
Looks good overall (but I haven't reviewed the UI parts yet), but I was wondering if we can't do this with Sentry directly.

**Comment by La0** [comment]
Advantages over Sentry:

1. Flexibility, i can build the frontend i want, displaying each ticket state.
2. Statistics, i can display precise statistics per error source over the logged period (a week)
3. Clustering, Sentry is not so good in the mercurial cases where the output varies

**Comment by marco-c** [comment]
> Flexibility, i can build the frontend i want, displaying each ticket state.

True, but if Sentry's one is good enough then we have less code to maintain. What features do you need that Sentry doesn't provide?
Also, Sentry has a lot of additional features that at some point you might want (e.g. filtering, searching and so on), and then you'd have to reimplement them too.

> Statistics, i can display precise statistics per error source over the logged period (a week)

Can't we do the same with Sentry? You can see the number of issues over the past 24h or the past 14 days.

> Clustering, Sentry is not so good in the mercurial cases where the output varies

By logging the error without the variable output you can cluster them together, e.g. https://github.com/mozilla/release-services/blob/bd1816ae4bc47805213d97fe78d2cf8bd6ae0472/src/codecoverage/bot/code_coverage_bot/grcov.py#L36 means that Sentry is clustering all of the errors that happen here as "Error while running grcov" (https://sentry.prod.mozaws.net/operations/mozilla-releng-services/issues/4368588/), but you can still see them individually if you want.

Let's discuss on IRC. We can merge this in the meantime if you want, and then figure out later if we want to keep it.

File: src/staticanalysis/bot/static_analysis_bot/revisions.py [hunk 2]
```diff
@@ -178,6 +179,9 @@ def load(self, repo):
             # Use base revision from top diff
             hg_base = self.diff['baseRevision']
 
+        if hg_base is None:
+            raise AnalysisException('phabricator', 'Missing base revision')
```
Comment (marco-c) [inline src/staticanalysis/bot/static_analysis_bot/revisions.py]: We should not fail in this case, but try to apply the patch to latest mozilla-central. Maybe just log a warning here

File: src/staticanalysis/bot/static_analysis_bot/revisions.py [hunk 3]
```diff
@@ -193,17 +197,20 @@ def load(self, repo):
                 clean=True,
             )
         except hglib.error.CommandError as e:
-            logger.warning('Failed to update to base revision', revision=hg_base, error=e)
+            raise AnalysisException('mercurial', 'Failed to update to base revision {}'.format(hg_base))
```
Comment (marco-c) [inline src/staticanalysis/bot/static_analysis_bot/revisions.py]: Same here

**Pull request merged**